File: unique-paths-ii.py
class Solution(object):
    def uniquePathsWithObstacles(self, obstacleGrid):
        """
        :type obstacleGrid: List[List[int]]
        :rtype: int
        """
        m = len(obstacleGrid)
        n = len(obstacleGrid[0])

        paths = [ [ 0 for j in range(n) ] for i in range(m) ]
        paths[m - 1][n - 1] = 1 - obstacleGrid[m - 1][n - 1]
        for i in reversed(range(m - 1)):
            if not obstacleGrid[i][n - 1]:
                paths[i][n - 1] = paths[i + 1][n - 1]
        for j in reversed(range(n - 1)):
            if not obstacleGrid[m - 1][j]:
                paths[m - 1][j] = paths[m - 1][j + 1]

        for i in reversed(range(m - 1)):
            for j in reversed(range(n - 1)):
                if not obstacleGrid[i][j]:
                    paths[i][j] = paths[i + 1][j] + paths[i][j + 1]

        return paths[0][0]

# assert Solution().uniquePathsWithObstacles([[0,0], [0,0]]) == 2

File: unique-paths-ii.js
/**
 * @param {number[][]} obstacleGrid
 * @return {number}
 */
var uniquePathsWithObstacles = function(obstacleGrid) {
  const m = obstacleGrid.length
  const n = obstacleGrid[0].length

  // initialize paths
  const paths = new Array(m)
  for (let i=0; i<m; i++) {
    paths[i] = new Array(n).fill(0)
  }
  paths[0][0] = 1 - obstacleGrid[0][0]
  for (let i=1; i<m; i++) {
    paths[i][0] = paths[i-1][0] * (1 - obstacleGrid[i][0])
  }
  for (let j=1; j<n; j++) {
    paths[0][j] = paths[0][j-1] * (1 - obstacleGrid[0][j])
  }

  // dynamic programming
  for (let i=1; i<m; i++) {
    for (let j=1; j<n; j++) {
      if (obstacleGrid[i][j] === 0) {
        paths[i][j] = paths[i-1][j] + paths[i][j-1]
      }
    }
  }

  return paths[m-1][n-1]
};
